File: java/src/test/java/fun/pierre/lists/P07Test.java
package fun.pierre.lists;

import org.junit.Test;

import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class P07Test {
    @Test
    public void shoulTestNestedIntList() {
        assertThat(P07.flatten(asList(1, 2, 3, 2, 1)), is(asList(1, 2, 3, 2, 1)));
        assertThat(P07.flatten(asList(
                asList(1, 2),
                asList(3, 2, 1)
        )), is(asList(1, 2, 3, 2, 1)));
    }

    @Test
    public void shoulTestNestedCharList() {
        assertThat(P07.flatten(asList('a', 'b', 'b', 'a')), is(asList('a', 'b', 'b', 'a')));
        assertThat(P07.flatten(asList(
                asList('a', 'b'),
                asList('b', 'a')
        )), is(asList('a', 'b', 'b', 'a')));
    }

    @Test
    public void shoulTestNestedIntListStream() {
        assertThat(P07.flatten_stream(asList(1, 2, 3, 2, 1)), is(asList(1, 2, 3, 2, 1)));
        assertThat(P07.flatten_stream(asList(
                asList(1, 2),
                asList(3, 2, 1)
        )), is(asList(1, 2, 3, 2, 1)));
    }

    @Test
    public void shoulTestNestedCharListStream() {
        assertThat(P07.flatten_stream(asList('a', 'b', 'b', 'a')), is(asList('a', 'b', 'b', 'a')));
        assertThat(P07.flatten_stream(asList(
                asList('a', 'b'),
                asList('b', 'a')
        )), is(asList('a', 'b', 'b', 'a')));
    }

}

File: scala/src/test/scala/fun/pierre/lists/P07Test.scala
package fun.pierre.lists

import org.scalatest.{FunSpec, Matchers}

class P07Test extends FunSpec with Matchers {

  describe("P07 flatten Char List") {
    it("P06 palindrom char element") {
      P07.flatten(List("a", "b", "b", "a")) should be(List("a", "b", "b", "a"))
      P07.flatten(List(List("a", "b"), List("b", "a"))) should be(List("a", "b", "b", "a"))
    }

    it("P07 flatten Int List") {
      P07.flatten(List(1, 2, 3, 2, 1)) should be(List(1, 2, 3, 2, 1))
      P07.flatten(List(List(1, 2), List(3, 2, 1))) should be(List(1, 2, 3, 2, 1))
    }
  }
}
